source/Halibut.Tests/Support/LatestClientAndLatestServiceBuilder.cs

```diff
@@ -102,127 +80,71 @@ public static LatestClientAndLatestServiceBuilder ForServiceConnectionType(Servi
-        }
-        
-        public LatestClientAndLatestServiceBuilder WithHalibutTimeoutsAndLimits(HalibutTimeoutsAndLimits halibutTimeoutsAndLimits)
-        {
-            this.halibutTimeoutsAndLimits = halibutTimeoutsAndLimits;
+            clientBuilder.WithClientConnectionsObserver(connectionsObserver);
             return this;
         }
 
-        public LatestClientAndLatestServiceBuilder WithAsyncConventionsDisabled()
+        public LatestClientAndLatestServiceBuilder WithHalibutTimeoutsAndLimits(HalibutTimeoutsAndLimits halibutTimeoutsAndLimits)
         {
-            serviceFactoryBuilder = serviceFactoryBuilder.WithConventionVerificationDisabled();
+            clientBuilder.WithHalibutTimeoutsAndLimits(halibutTimeoutsAndLimits);
+            serviceBuilder.WithHalibutTimeoutsAndLimits(halibutTimeoutsAndLimits);
             return this;
         }
 
         public LatestClientAndLatestServiceBuilder WithServiceFactory(IServiceFactory serviceFactory)
         {
-            this.serviceFactory = serviceFactory;
+            serviceBuilder.WithServiceFactory(serviceFactory);
             return this;
         }
 
         public LatestClientAndLatestServiceBuilder WithAsyncService<TContract, TClientContract>(Func<TClientContract> implementation)
         {
-            serviceFactoryBuilder.WithService<TContract, TClientContract>(implementation);
-            
-            if (serviceFactory != null)
-            {
-                if (serviceFactory is DelegateServiceFactory delegateServiceFactory)
-                {
-                    delegateServiceFactory.Register<TContract, TClientContract>(implementation);
-                }
-                else
-                {
-                    throw new Exception("WithService can only be used with a custom ServiceFactory if it is a DelegateServiceFactory");
-                }
-            }
-
+            serviceBuilder.WithAsyncService<TContract, TClientContract>(implementation);
             return this;
         }
 
-        public LatestClientAndLatestServiceBuilder WithPortForwarding()
+        IClientAndServiceBuilder IClientAndServiceBuilder.WithPortForwarding(out Reference<PortForwarder> portForwarder, Func<int, PortForwarder> portForwarderFactory)
         {
-            return WithPortForwarding(port => PortForwarderUtil.ForwardingToLocalPort(port).Build());
+            return this.WithPortForwarding(out portForwarder, portForwarderFactory);
         }
 
-        IClientAndServiceBuilder IClientAndServiceBuilder.WithPortForwarding(Func<int, PortForwarder> portForwarderFactory)
+        public LatestClientAndLatestServiceBuilder WithPortForwarding(out Reference<PortForwarder> portForwarder, Func<int, PortForwarder> portForwarderFactory)
         {
-            return this.WithPortForwarding(portForwarderFactory);
-        }
+            // Based on ServiceConnectionType, the port forwarder will be created in either clientBuilder or serviceBuilder.
```

Comment:
I don't like this, so please let me know if there is a better way.

The issue is that we do not know which builder will end up making the port forwarder. So we have to register references for both, and then see later which one was the one that made it 😑